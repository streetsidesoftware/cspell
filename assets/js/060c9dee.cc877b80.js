"use strict";(self.webpackChunkwebsite=self.webpackChunkwebsite||[]).push([[94023],{46039:(e,t,n)=>{n.d(t,{R:()=>l,x:()=>c});var i=n(18225);const s={},r=i.createContext(s);function l(e){const t=i.useContext(r);return i.useMemo((function(){return"function"==typeof e?e(t):{...t,...e}}),[t,e])}function c(e){let t;return t=e.disableParentContext?"function"==typeof e.components?e.components(s):e.components||s:l(e.components),i.createElement(r.Provider,{value:t},e.children)}},87257:(e,t,n)=>{n.r(t),n.d(t,{assets:()=>a,contentTitle:()=>c,default:()=>p,frontMatter:()=>l,metadata:()=>i,toc:()=>o});const i=JSON.parse('{"id":"api/cspell-lib/functions/readRawSettings","title":"Function: readRawSettings()","description":"readRawSettings(filename, relativeTo?): Promise\\\\","source":"@site/docs/api/cspell-lib/functions/readRawSettings.md","sourceDirName":"api/cspell-lib/functions","slug":"/api/cspell-lib/functions/readRawSettings","permalink":"/docs/api/cspell-lib/functions/readRawSettings","draft":false,"unlisted":false,"editUrl":"https://github.com/streetsidesoftware/cspell/tree/main/website/docs/docs/api/cspell-lib/functions/readRawSettings.md","tags":[],"version":"current","frontMatter":{},"sidebar":"tutorialSidebar","previous":{"title":"Function: readFileSync()","permalink":"/docs/api/cspell-lib/functions/readFileSync"},"next":{"title":"Function: readSettings()","permalink":"/docs/api/cspell-lib/functions/readSettings"}}');var s=n(37557),r=n(46039);const l={},c="Function: readRawSettings()",a={},o=[{value:"Parameters",id:"parameters",level:2},{value:"filename",id:"filename",level:3},{value:"relativeTo?",id:"relativeto",level:3},{value:"Returns",id:"returns",level:2}];function d(e){const t={a:"a",blockquote:"blockquote",code:"code",h1:"h1",h2:"h2",h3:"h3",header:"header",p:"p",strong:"strong",...(0,r.R)(),...e.components};return(0,s.jsxs)(s.Fragment,{children:[(0,s.jsx)(t.header,{children:(0,s.jsx)(t.h1,{id:"function-readrawsettings",children:"Function: readRawSettings()"})}),"\n",(0,s.jsxs)(t.blockquote,{children:["\n",(0,s.jsxs)(t.p,{children:[(0,s.jsx)(t.strong,{children:"readRawSettings"}),"(",(0,s.jsx)(t.code,{children:"filename"}),", ",(0,s.jsx)(t.code,{children:"relativeTo"}),"?): ",(0,s.jsx)(t.code,{children:"Promise"}),"<",(0,s.jsx)(t.a,{href:"/docs/api/cspell-lib/interfaces/CSpellSettingsWithSourceTrace",children:(0,s.jsx)(t.code,{children:"CSpellSettingsWithSourceTrace"})}),">"]}),"\n"]}),"\n",(0,s.jsxs)(t.p,{children:["Defined in: ",(0,s.jsx)(t.a,{href:"https://github.com/streetsidesoftware/cspell/blob/3fc3ef2ea9def847ef20287c6238ff423e6493eb/packages/cspell-lib/src/lib/Settings/Controller/configLoader/defaultConfigLoader.ts#L99",children:"packages/cspell-lib/src/lib/Settings/Controller/configLoader/defaultConfigLoader.ts:99"})]}),"\n",(0,s.jsx)(t.h2,{id:"parameters",children:"Parameters"}),"\n",(0,s.jsx)(t.h3,{id:"filename",children:"filename"}),"\n",(0,s.jsxs)(t.p,{children:[(0,s.jsx)(t.code,{children:"string"})," | ",(0,s.jsx)(t.code,{children:"URL"})]}),"\n",(0,s.jsx)(t.h3,{id:"relativeto",children:"relativeTo?"}),"\n",(0,s.jsxs)(t.p,{children:[(0,s.jsx)(t.code,{children:"string"})," | ",(0,s.jsx)(t.code,{children:"URL"})]}),"\n",(0,s.jsx)(t.h2,{id:"returns",children:"Returns"}),"\n",(0,s.jsxs)(t.p,{children:[(0,s.jsx)(t.code,{children:"Promise"}),"<",(0,s.jsx)(t.a,{href:"/docs/api/cspell-lib/interfaces/CSpellSettingsWithSourceTrace",children:(0,s.jsx)(t.code,{children:"CSpellSettingsWithSourceTrace"})}),">"]})]})}function p(e={}){const{wrapper:t}={...(0,r.R)(),...e.components};return t?(0,s.jsx)(t,{...e,children:(0,s.jsx)(d,{...e})}):d(e)}}}]);