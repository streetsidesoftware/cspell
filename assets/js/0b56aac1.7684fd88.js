"use strict";(self.webpackChunkwebsite=self.webpackChunkwebsite||[]).push([[16718],{17322:(e,s,i)=>{i.r(s),i.d(s,{assets:()=>o,contentTitle:()=>a,default:()=>p,frontMatter:()=>r,metadata:()=>l,toc:()=>c});const l=JSON.parse('{"id":"api/cspell-lib/variables/ConfigFields","title":"Variable: ConfigFields","description":"const ConfigFields: CSpellUserSettingsFields","source":"@site/docs/api/cspell-lib/variables/ConfigFields.md","sourceDirName":"api/cspell-lib/variables","slug":"/api/cspell-lib/variables/ConfigFields","permalink":"/docs/api/cspell-lib/variables/ConfigFields","draft":false,"unlisted":false,"editUrl":"https://github.com/streetsidesoftware/cspell/tree/main/website/docs/docs/api/cspell-lib/variables/ConfigFields.md","tags":[],"version":"current","frontMatter":{},"sidebar":"tutorialSidebar","previous":{"title":"Type Alias: ~~VersionLegacy~~","permalink":"/docs/api/cspell-lib/type-aliases/VersionLegacy"},"next":{"title":"Variable: ENV\\\\_CSPELL\\\\_GLOB\\\\_ROOT","permalink":"/docs/api/cspell-lib/variables/ENV_CSPELL_GLOB_ROOT"}}');var t=i(37557),n=i(46039);const r={},a="Variable: ConfigFields",o={},c=[];function d(e){const s={a:"a",blockquote:"blockquote",code:"code",h1:"h1",header:"header",p:"p",strong:"strong",...(0,n.R)(),...e.components};return(0,t.jsxs)(t.Fragment,{children:[(0,t.jsx)(s.header,{children:(0,t.jsx)(s.h1,{id:"variable-configfields",children:"Variable: ConfigFields"})}),"\n",(0,t.jsxs)(s.blockquote,{children:["\n",(0,t.jsxs)(s.p,{children:[(0,t.jsx)(s.code,{children:"const"})," ",(0,t.jsx)(s.strong,{children:"ConfigFields"}),": ",(0,t.jsx)(s.a,{href:"/docs/api/cspell-lib/type-aliases/CSpellUserSettingsFields",children:(0,t.jsx)(s.code,{children:"CSpellUserSettingsFields"})})]}),"\n"]}),"\n",(0,t.jsx)(s.p,{children:"Defined in: packages/cspell-types/dist/index.d.mts:1566"})]})}function p(e={}){const{wrapper:s}={...(0,n.R)(),...e.components};return s?(0,t.jsx)(s,{...e,children:(0,t.jsx)(d,{...e})}):d(e)}},46039:(e,s,i)=>{i.d(s,{R:()=>r,x:()=>a});var l=i(18225);const t={},n=l.createContext(t);function r(e){const s=l.useContext(n);return l.useMemo((function(){return"function"==typeof e?e(s):{...s,...e}}),[s,e])}function a(e){let s;return s=e.disableParentContext?"function"==typeof e.components?e.components(t):e.components||t:r(e.components),l.createElement(n.Provider,{value:s},e.children)}}}]);