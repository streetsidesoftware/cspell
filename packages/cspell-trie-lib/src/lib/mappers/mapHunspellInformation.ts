import { opFilter, opFlatten, opMap, pipeSync as pipe } from '@cspell/cspell-pipe';
import type { HunspellCosts, HunspellInformation } from '../models/DictionaryInformation';
import { Locale } from '../models/locale';
import type { SuggestionCostMapDef } from '../models/suggestionCostsDef';
import { caseForms } from '../utils/text';
import { isDefined, unique as uniqueU } from '../utils/util';
import { mapHunspellCosts } from './mapCosts';
import {
    calcCostsForAccentedLetters,
    calcFirstCharacterReplace,
    parseAlphabet,
    splitMap,
} from './mapToSuggestionCostDef';

interface Costs extends Required<HunspellCosts> {
    locale?: string[] | undefined;
}

export function hunspellInformationToSuggestionCostDef(
    hunInfo: HunspellInformation,
    locales: Locale[] | undefined
): SuggestionCostMapDef[] {
    const costs = calcCosts(hunInfo.costs, locales);

    const operations = [
        affKey,
        affKeyCaps,
        affMap,
        affMapAccents,
        affMapCaps,
        affNoTry,
        affRepConv,
        affTry,
        affTryAccents,
        affTryFirstCharacterReplace,
    ];

    function parseAff(aff: string, costs: Costs) {
        // cspell:ignore OCONV
        const regSupportedAff = /^(?:MAP|KEY|TRY|NO-TRY|ICONV|OCONV|REP)\s/;
        const rejectAff = /^(?:MAP|KEY|TRY|ICONV|OCONV|REP)\s+\d+$/;

        const lines = aff
            .split('\n')
            .map((a) => a.replace(/#.*/, ''))
            .map((a) => a.trim())
            .filter((a) => regSupportedAff.test(a))
            .filter((a) => !rejectAff.test(a));

        const defs = pipe(
            lines,
            opMap((line) =>
                pipe(
                    operations,
                    opMap((fn) => fn(line, costs)),
                    opMap(asArrayOf),
                    opFlatten()
                )
            ),
            opFlatten(),
            opFilter(isDefined)
        );

        return [...defs];
    }

    return parseAff(hunInfo.aff, costs);
}

function calcCosts(costs: HunspellCosts = {}, locale?: Locale[] | undefined): Costs {
    const useLocale = locale?.length ? locale.map((loc) => loc.locale) : undefined;

    const hunCosts = mapHunspellCosts(costs);

    const c: Costs = {
        ...hunCosts,
        locale: useLocale,
    };
    return c;
}

const regExpMap = /^(?:MAP)\s+(\S+)$/;

function affMap(line: string, costs: Costs): SuggestionCostMapDef | undefined {
    const m = line.match(regExpMap);
    if (!m) return undefined;

    const map = m[1];
    const cost = costs.mapCost;

    return {
        map,
        replace: cost,
        swap: cost,
    };
}

const regExpTry = /^(?:TRY)\s+(\S+)$/;

function affTry(line: string, costs: Costs): SuggestionCostMapDef[] | undefined {
    const m = line.match(regExpTry);
    if (!m) return undefined;

    const cost = costs.tryCharCost;
    const tryChars = m[1];
    const characters = tryChars;

    return parseAlphabet(
        {
            characters,
            cost,
        },
        costs.locale,
        costs
    );
}

function affTryFirstCharacterReplace(line: string, costs: Costs): SuggestionCostMapDef | undefined {
    const m = line.match(regExpTry);
    if (!m) return undefined;

    const characters = m[1];
    // Make it a bit cheaper so it will match
    const cost = costs.tryCharCost;

    return calcFirstCharacterReplace(
        {
            characters,
            cost,
        },
        costs
    );
}

const regExpNoTry = /^NO-TRY\s+(\S+)$/;

function affNoTry(line: string, costs: Costs): SuggestionCostMapDef | undefined {
    const m = line.match(regExpNoTry);
    if (!m) return undefined;

    const map = m[1];

    return {
        map,
        insDel: Math.max(costs.nonAlphabetCosts - costs.tryCharCost, 0),
        penalty: costs.nonAlphabetCosts + costs.tryCharCost,
    };
}

// cspell:ignore conv
const regExpRepConv = /^(?:REP|(?:I|O)CONV)\s+(\S+)\s+(\S+)$/;

function affRepConv(line: string, costs: Costs): SuggestionCostMapDef | undefined {
    const m = line.match(regExpRepConv);
    if (!m) return undefined;

    const cost = line.startsWith('REP') ? costs.replaceCosts : costs.ioConvertCost;
    const from = m[1];
    let into = m[2];
    into = into.replace(/^0$/, '');

    if (from.startsWith('^') && !into.startsWith('^')) {
        into = '^' + into;
    }

    if (from.endsWith('$') && !into.endsWith('$')) {
        into = into + '$';
    }

    return {
        map: joinLetters([from, into]),
        replace: cost,
    };
}

const regExpKey = /^(?:KEY)\s+(\S+)$/;

function affKey(line: string, costs: Costs): SuggestionCostMapDef | undefined {
    const m = line.match(regExpKey);
    if (!m) return undefined;

    const kbd = m[1];

    const pairs = [...splitMap(kbd)]
        .map(reducer((p, v) => ({ a: p.b, b: v }), { a: '|', b: '|' }))
        .filter((ab) => ab.a !== '|' && ab.b !== '|')
        .map(({ a, b }) => joinLetters([a, b]));

    const pairsUpper = pairs.map((p) => p.toLocaleUpperCase(costs.locale));

    const map = uniqueU(pairs.concat(pairsUpper)).join('|');
    const cost = costs.keyboardCost;

    return {
        map,
        replace: cost,
        swap: cost,
    };
}

function affKeyCaps(line: string, costs: Costs): SuggestionCostMapDef | undefined {
    const m = line.match(regExpKey);
    if (!m) return undefined;
    return parseCaps(m[1], costs);
}

function affMapCaps(line: string, costs: Costs): SuggestionCostMapDef | undefined {
    const m = line.match(regExpMap);
    if (!m) return undefined;
    return parseCaps(m[1], costs);
}

function affTryAccents(line: string, costs: Costs): SuggestionCostMapDef[] | undefined {
    const m = line.match(regExpTry);
    if (!m) return undefined;
    return calcCostsForAccentedLetters(m[1], costs.locale, costs);
}

function affMapAccents(line: string, costs: Costs): SuggestionCostMapDef[] | undefined {
    const m = line.match(regExpMap);
    if (!m) return undefined;
    return calcCostsForAccentedLetters(m[1], costs.locale, costs);
}

function parseCaps(value: string, costs: Costs): SuggestionCostMapDef | undefined {
    const locale = costs.locale;
    const letters = [...splitMap(value)].filter((a) => a !== '|');
    const withCases = letters
        .map((s) => caseForms(s, locale))
        .filter((forms) => forms.length > 1)
        .map(joinLetters);

    const map = uniqueU(withCases).join('|');
    const cost = costs.capsCosts;

    if (!map) return undefined;

    return {
        map,
        replace: cost,
    };
}

/**
 * Bring letters / strings together.
 * - `['a', 'b'] => 'ab'`
 * - `['a', 'bc'] => 'a(bc)'`
 * @param letters - letters to join
 */
export function joinLetters(letters: string[]): string {
    const v = [...letters];
    return v.map((a) => (a.length > 1 || !a.length ? `(${a})` : a)).join('');
}

function reducer<T, U = T>(fn: (acc: U, val: T, i: number) => U, initialVal: U) {
    let acc = initialVal;
    return (val: T, i: number) => (acc = fn(acc, val, i));
}

function asArrayOf<T>(v: T | T[]): T[] {
    return Array.isArray(v) ? v : [v];
}

export const __testing__ = {
    affKey,
    affKeyCaps,
    affMap,
    affMapAccents,
    affMapCaps,
    affNoTry,
    affRepConv,
    affTry,
    affTryAccents,
    affTryFirstCharacterReplace,
    calcCosts,
    split: splitMap,
};
