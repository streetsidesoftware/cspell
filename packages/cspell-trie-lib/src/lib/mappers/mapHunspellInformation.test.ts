import { describe, expect, test } from 'vitest';

import type { HunspellCosts, HunspellInformation } from '../models/DictionaryInformation.js';
import { __testing__, hunspellInformationToSuggestionCostDef } from './mapHunspellInformation.js';

// cspell:ignore conv OCONV
const {
    affMap,
    affRepConv,
    affNoTry,
    affTry,
    affKey,
    affTryFirstCharacterReplace,
    calcCosts,
    affMapCaps,
    affKeyCaps,
    affTryAccents,
    affMapAccents,
} = __testing__;

const sampleAff = `
# comment.
# cspell:ignore eéèëê iíìïî
TRY abc

MAP 3
MAP eéèëê
MAP aáà
MAP iíìïî

OCONV 2
OCONV ĳ ij
OCONV Ĳ IJ

ICONV 9
ICONV áá aa
ICONV éé ee
ICONV íé ie
ICONV óó oo
ICONV úú uu
ICONV óé oe
ICONV ’ '
ICONV ij ĳ
ICONV IJ Ĳ

REP ^ß ss

NO-TRY -1234567890

# cspell:ignore asdfghjkl qwertyuiop zxcvbnm
KEY qwertyuiop|asdfghjkl|zxcvbnm
`;

describe('mapHunspellInformation', () => {
    // cspell:ignore aàâä
    test.each`
        line                  | costs             | expected
        ${''}                 | ${{}}             | ${undefined}
        ${'MAP aàâäAÀÂÄ'}     | ${{}}             | ${{ map: 'aàâäAÀÂÄ', replace: 25, swap: 25 }}
        ${'MAP 😁😀😊😂🤣😬'} | ${{}}             | ${{ map: '😁😀😊😂🤣😬', replace: 25, swap: 25 }}
        ${'MAP aàâäAÀÂÄ'}     | ${{ mapCost: 1 }} | ${{ map: 'aàâäAÀÂÄ', replace: 1, swap: 1 }}
        ${'MAP ß(ss)'}        | ${{}}             | ${{ map: 'ß(ss)', replace: 25, swap: 25 }}
    `('affMap "$line" $costs', ({ line, costs, expected }) => {
        expect(affMap(line, calcCosts(costs))).toEqual(expected);
    });

    test.each`
        line                  | costs                  | expected
        ${''}                 | ${{}}                  | ${undefined}
        ${'MAP aàâäAÀÂÄ'}     | ${{}}                  | ${{ map: 'aA|àÀ|âÂ|äÄ|Aa|Àà|Ââ|Ää', replace: 1 }}
        ${'MAP 😁😀😊😂🤣😬'} | ${{}}                  | ${undefined}
        ${'MAP aàâäAÀÂÄ'}     | ${c({ capsCosts: 2 })} | ${{ map: 'aA|àÀ|âÂ|äÄ|Aa|Àà|Ââ|Ää', replace: 2 }}
        ${'MAP ß(ss)'}        | ${{}}                  | ${{ map: 'ß(SS)(ss)|(ss)(SS)', replace: 1 }}
    `('affMapCaps "$line" $costs', ({ line, costs, expected }) => {
        expect(affMapCaps(line, calcCosts(costs))).toEqual(expected);
    });

    test.each`
        line                  | costs                    | expected
        ${''}                 | ${{}}                    | ${undefined}
        ${'MAP 😁😀😊😂🤣😬'} | ${{}}                    | ${[]}
        ${'MAP aàâäAÀÂÄ'}     | ${{}}                    | ${[{ map: 'a(à)à|A(À)À|a(â)â|A(Â)Â|a(ä)ä|A(Ä)Ä', replace: 1 }, { map: '(à)à|(À)À|(â)â|(Â)Â|(ä)ä|(Ä)Ä|(À)À|(à)à|(Â)Â|(â)â|(Ä)Ä|(ä)ä', replace: 0 }]}
        ${'MAP aàâäAÀÂÄ'}     | ${c({ accentCosts: 2 })} | ${[{ map: 'a(à)à|A(À)À|a(â)â|A(Â)Â|a(ä)ä|A(Ä)Ä', replace: 2 }, { map: '(à)à|(À)À|(â)â|(Â)Â|(ä)ä|(Ä)Ä|(À)À|(à)à|(Â)Â|(â)â|(Ä)Ä|(ä)ä', replace: 0 }]}
        ${'MAP ß(ss)'}        | ${{}}                    | ${[]}
    `('affMapCaps "$line" $costs', ({ line, costs, expected }) => {
        expect(affMapAccents(line, calcCosts(costs))).toEqual(expected);
    });

    test.each`
        line                 | costs                            | expected
        ${''}                | ${{}}                            | ${undefined}
        ${'MAP aàâäAÀÂÄ'}    | ${{}}                            | ${undefined}
        ${'NO-TRY aàâäAÀÂÄ'} | ${{}}                            | ${undefined}
        ${'TRY abc'}         | ${c({ mapCost: 1 })}             | ${[{ map: 'ABCabc', replace: 100, insDel: 100, swap: 100 }, { map: 'Aa|Bb|Cc', replace: 1 }]}
        ${'TRY abc'}         | ${c({ tryCharCost: 90 })}        | ${[{ map: 'ABCabc', replace: 90, insDel: 90, swap: 90 }, { map: 'Aa|Bb|Cc', replace: 1 }]}
        ${'TRY abc'}         | ${c({ firstLetterPenalty: 10 })} | ${[{ map: 'ABCabc', replace: 100, insDel: 100, swap: 100 }, { map: 'Aa|Bb|Cc', replace: 1 }]}
    `('affTry "$line" $costs', ({ line, costs, expected }) => {
        expect(affTry(line, calcCosts(costs))).toEqual(expected);
    });

    test.each`
        line              | costs                            | expected
        ${''}             | ${{}}                            | ${undefined}
        ${'MAP aàâäAÀÂÄ'} | ${{}}                            | ${undefined}
        ${'TRY aàâäAÀÂÄ'} | ${{}}                            | ${undefined}
        ${'NO-TRY abc'}   | ${c({ mapCost: 1 })}             | ${{ map: 'abc', insDel: 10, penalty: 210 }}
        ${'NO-TRY abc'}   | ${c({ tryCharCost: 90 })}        | ${{ map: 'abc', insDel: 20, penalty: 200 }}
        ${'NO-TRY abc'}   | ${c({ firstLetterPenalty: 10 })} | ${{ map: 'abc', insDel: 10, penalty: 210 }}
    `('affNoTry "$line" $costs', ({ line, costs, expected }) => {
        expect(affNoTry(line, calcCosts(costs))).toEqual(expected);
    });

    test.each`
        line              | costs                            | expected
        ${''}             | ${{}}                            | ${undefined}
        ${'MAP aàâäAÀÂÄ'} | ${{}}                            | ${undefined}
        ${'TRY abc'}      | ${c({ mapCost: 1 })}             | ${{ map: '(^a)(^b)(^c)(^)', replace: 96, penalty: 8 }}
        ${'TRY abc'}      | ${c({ tryCharCost: 90 })}        | ${{ map: '(^a)(^b)(^c)(^)', replace: 86, penalty: 8 }}
        ${'TRY abc'}      | ${c({ firstLetterPenalty: 10 })} | ${{ map: '(^a)(^b)(^c)(^)', replace: 90, penalty: 20 }}
    `('affTryFirstCharacterReplace "$line" $costs', ({ line, costs, expected }) => {
        expect(affTryFirstCharacterReplace(line, calcCosts(costs))).toEqual(expected);
    });

    // cspell:ignore qwer zxcv
    test.each`
        line                          | costs                      | expected
        ${''}                         | ${{}}                      | ${undefined}
        ${'MAP aàâäAÀÂÄ'}             | ${{}}                      | ${undefined}
        ${'KEY qwer|asdf|zxcv'}       | ${c({ mapCost: 1 })}       | ${{ map: 'qw|we|er|as|sd|df|zx|xc|cv|QW|WE|ER|AS|SD|DF|ZX|XC|CV', replace: 99, swap: 99 }}
        ${'KEY qwer|a|asdf|zxcv'}     | ${c({ keyboardCost: 74 })} | ${{ map: 'qw|we|er|as|sd|df|zx|xc|cv|QW|WE|ER|AS|SD|DF|ZX|XC|CV', replace: 74, swap: 74 }}
        ${'KEY qwer春😁|a|asdf|zxcv'} | ${c()}                     | ${{ map: 'qw|we|er|r春|春(😁)|as|sd|df|zx|xc|cv|QW|WE|ER|R春|AS|SD|DF|ZX|XC|CV', replace: 99, swap: 99 }}
        ${'KEY a😁b'}                 | ${c()}                     | ${{ map: 'a(😁)|(😁)b|A(😁)|(😁)B', replace: 99, swap: 99 }}
    `('affKey "$line" $costs', ({ line, costs, expected }) => {
        expect(affKey(line, calcCosts(costs))).toEqual(expected);
    });

    test.each`
        line                          | costs                      | expected
        ${''}                         | ${{}}                      | ${undefined}
        ${'MAP aàâäAÀÂÄ'}             | ${{}}                      | ${undefined}
        ${'KEY qwer|a|asdf|zxcv'}     | ${c({ keyboardCost: 74 })} | ${{ map: 'qQ|wW|eE|rR|aA|sS|dD|fF|zZ|xX|cC|vV', replace: 1 }}
        ${'KEY qwer|a|asdf|zxcv'}     | ${c({ capsCosts: 2 })}     | ${{ map: 'qQ|wW|eE|rR|aA|sS|dD|fF|zZ|xX|cC|vV', replace: 2 }}
        ${'KEY qwer春😁|a|asdf|zxcv'} | ${c()}                     | ${{ map: 'qQ|wW|eE|rR|aA|sS|dD|fF|zZ|xX|cC|vV', replace: 1 }}
        ${'KEY a😁b'}                 | ${c()}                     | ${{ map: 'aA|bB', replace: 1 }}
    `('affKeyCaps "$line" $costs', ({ line, costs, expected }) => {
        expect(affKeyCaps(line, calcCosts(costs))).toEqual(expected);
    });

    // cspell:ignore qwér aasdfzxcv abcdéefghi
    test.each`
        line                    | costs                      | expected
        ${''}                   | ${{}}                      | ${undefined}
        ${'MAP aàâäAÀÂÄ'}       | ${{}}                      | ${undefined}
        ${'TRY abcdéefghi'}     | ${c({ keyboardCost: 74 })} | ${[{ map: 'e(é)é|E(É)É', replace: 1 }, { map: '(é)é|(É)É', replace: 0 }]}
        ${'TRY abcdéefghi'}     | ${c({ accentCosts: 33 })}  | ${[{ map: 'e(é)é|E(É)É', replace: 33 }, { map: '(é)é|(É)É', replace: 0 }]}
        ${'TRY 春😁|aasdfzxcv'} | ${c()}                     | ${[]}
        ${'TRY a😁b'}           | ${c()}                     | ${[]}
    `('affTryAccents "$line" $costs', ({ line, costs, expected }) => {
        expect(affTryAccents(line, calcCosts(costs))).toEqual(expected);
    });

    test.each`
        line              | costs                    | expected
        ${''}             | ${{}}                    | ${undefined}
        ${'REP o oo'}     | ${{}}                    | ${{ map: 'o(oo)', replace: 75 }}
        ${'REP ^a A'}     | ${{ mapCost: 1 }}        | ${{ map: '(^a)(^A)', replace: 75 }}
        ${'REP $ en$'}    | ${{ replaceCosts: 55 }}  | ${{ map: '$(en$)', replace: 55 }}
        ${'REP ^af$ aff'} | ${{}}                    | ${{ map: '(^af$)(^aff$)', replace: 75 }}
        ${'REP ß ss'}     | ${{}}                    | ${{ map: 'ß(ss)', replace: 75 }}
        ${'REP ß 0'}      | ${{}}                    | ${{ map: 'ß()', replace: 75 }}
        ${'REP 25'}       | ${{}}                    | ${undefined}
        ${'ICONV 25'}     | ${{}}                    | ${undefined}
        ${'OCONV 25'}     | ${{}}                    | ${undefined}
        ${'ICONV áá aa'}  | ${{}}                    | ${{ map: '(áá)(aa)', replace: 30 }}
        ${'OCONV ss ß'}   | ${{ replaceCosts: 55 }}  | ${{ map: '(ss)ß', replace: 30 }}
        ${'OCONV ss ß'}   | ${{ ioConvertCost: 25 }} | ${{ map: '(ss)ß', replace: 25 }}
    `('affRepConv "$line" $costs', ({ line, costs, expected }) => {
        expect(affRepConv(line, calcCosts(costs))).toEqual(expected);
    });

    const AFFs: [string, HunspellCosts | undefined][] = [
        ['', undefined],
        [
            `
            TRY abc
            `,
            undefined,
        ],
        [sampleAff, undefined],
    ];

    test.each(AFFs)('hunspellInformationToSuggestionCostDef <%s>', (aff, costs) => {
        const info: HunspellInformation = { aff };
        if (costs) {
            info.costs = costs;
        }
        const defs = hunspellInformationToSuggestionCostDef(info, undefined);
        expect(defs).toMatchSnapshot();
    });
});

function c(hc: HunspellCosts = {}): HunspellCosts {
    return hc;
}
