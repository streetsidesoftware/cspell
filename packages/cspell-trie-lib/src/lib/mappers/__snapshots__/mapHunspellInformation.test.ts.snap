// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapHunspellInformation hunspellInformationToSuggestionCostDef <
            TRY abc
            > 1`] = `
Array [
  Object {
    "insDel": 100,
    "map": "ABCabc",
    "replace": 100,
    "swap": 100,
  },
  Object {
    "map": "Aa|Bb|Cc",
    "replace": 1,
  },
  Object {
    "map": "(^a)(^b)(^c)(^)",
    "penalty": 8,
    "replace": 96,
  },
]
`;

exports[`mapHunspellInformation hunspellInformationToSuggestionCostDef <
# comment.
# cspell:ignore eéèëê iíìïî
TRY abc

MAP 3
MAP eéèëê
MAP aáà
MAP iíìïî

OCONV 2
OCONV ĳ ij
OCONV Ĳ IJ

ICONV 9
ICONV áá aa
ICONV éé ee
ICONV íé ie
ICONV óó oo
ICONV úú uu
ICONV óé oe
ICONV ’ '
ICONV ij ĳ
ICONV IJ Ĳ

REP ^ß ss

NO-TRY -1234567890

# cspell:ignore asdfghjkl qwertyuiop zxcvbnm
KEY qwertyuiop|asdfghjkl|zxcvbnm
> 1`] = `
Array [
  Object {
    "insDel": 100,
    "map": "ABCabc",
    "replace": 100,
    "swap": 100,
  },
  Object {
    "map": "Aa|Bb|Cc",
    "replace": 1,
  },
  Object {
    "map": "(^a)(^b)(^c)(^)",
    "penalty": 8,
    "replace": 96,
  },
  Object {
    "map": "eéèëê",
    "replace": 25,
    "swap": 25,
  },
  Object {
    "map": "e(é)é|E(É)É|e(è)è|E(È)È|e(ë)ë|E(Ë)Ë|e(ê)ê|E(Ê)Ê",
    "replace": 1,
  },
  Object {
    "map": "(é)é|(É)É|(è)è|(È)È|(ë)ë|(Ë)Ë|(ê)ê|(Ê)Ê",
    "replace": 0,
  },
  Object {
    "map": "eE|éÉ|èÈ|ëË|êÊ",
    "replace": 1,
  },
  Object {
    "map": "aáà",
    "replace": 25,
    "swap": 25,
  },
  Object {
    "map": "a(á)á|A(Á)Á|a(à)à|A(À)À",
    "replace": 1,
  },
  Object {
    "map": "(á)á|(Á)Á|(à)à|(À)À",
    "replace": 0,
  },
  Object {
    "map": "aA|áÁ|àÀ",
    "replace": 1,
  },
  Object {
    "map": "iíìïî",
    "replace": 25,
    "swap": 25,
  },
  Object {
    "map": "i(í)í|I(Í)Í|i(ì)ì|I(Ì)Ì|i(ï)ï|I(Ï)Ï|i(î)î|I(Î)Î",
    "replace": 1,
  },
  Object {
    "map": "(í)í|(Í)Í|(ì)ì|(Ì)Ì|(ï)ï|(Ï)Ï|(î)î|(Î)Î",
    "replace": 0,
  },
  Object {
    "map": "iI|íÍ|ìÌ|ïÏ|îÎ",
    "replace": 1,
  },
  Object {
    "map": "ĳ(ij)",
    "replace": 30,
  },
  Object {
    "map": "Ĳ(IJ)",
    "replace": 30,
  },
  Object {
    "map": "(áá)(aa)",
    "replace": 30,
  },
  Object {
    "map": "(éé)(ee)",
    "replace": 30,
  },
  Object {
    "map": "(íé)(ie)",
    "replace": 30,
  },
  Object {
    "map": "(óó)(oo)",
    "replace": 30,
  },
  Object {
    "map": "(úú)(uu)",
    "replace": 30,
  },
  Object {
    "map": "(óé)(oe)",
    "replace": 30,
  },
  Object {
    "map": "’'",
    "replace": 30,
  },
  Object {
    "map": "(ij)ĳ",
    "replace": 30,
  },
  Object {
    "map": "(IJ)Ĳ",
    "replace": 30,
  },
  Object {
    "map": "(^ß)(^ss)",
    "replace": 75,
  },
  Object {
    "insDel": 10,
    "map": "-1234567890",
    "penalty": 210,
  },
  Object {
    "map": "qw|we|er|rt|ty|yu|ui|io|op|as|sd|df|fg|gh|hj|jk|kl|zx|xc|cv|vb|bn|nm|QW|WE|ER|RT|TY|YU|UI|IO|OP|AS|SD|DF|FG|GH|HJ|JK|KL|ZX|XC|CV|VB|BN|NM",
    "replace": 99,
    "swap": 99,
  },
  Object {
    "map": "qQ|wW|eE|rR|tT|yY|uU|iI|oO|pP|aA|sS|dD|fF|gG|hH|jJ|kK|lL|zZ|xX|cC|vV|bB|nN|mM",
    "replace": 1,
  },
]
`;

exports[`mapHunspellInformation hunspellInformationToSuggestionCostDef <> 1`] = `Array []`;
